### PR TITLE
RSDK-628 - allow configurable kinematics for fake arm models

### PR DESCRIPTION
**Pull request by raybjork**
Currently there is no way to specify what kinematic model a fake arm which makes it difficult to test motion planning for arms that we are not connected to.  Additionally, the fake and fakeIK models both exist in the codebase which is confusing. This PR addresses both these concerns by wrapping fakeIK model functionality into the fake model, and further makes the kinematic model for the fake model configurable.

### CODE REVIEW

File: robot/framesystem/service_test.go [hunk 1]
```diff
@@ -145,19 +146,19 @@ func TestFrameSystemFromConfig(t *testing.T) {
 
 	// There is a point at (1500, 500, 1300) in the world referenceframe. See if it transforms correctly in each referenceframe.
 	worldPose := referenceframe.NewPoseInFrame(referenceframe.World, spatialmath.NewPoseFromPoint(r3.Vector{1500, 500, 1300}))
-	armPt := r3.Vector{0, 0, 500}
+	armPt := r3.Vector{500, 0, 0}
```
Comment (spieswl) [inline robot/framesystem/service_test.go:149]: Is there a place we can check where these values are coming from?

File: services/armremotecontrol/builtin/builtin_test.go [hunk 2]
```diff
@@ -65,7 +66,10 @@ func TestArmRemoteControl(t *testing.T) {
 		case input.Subtype:
 			return fakeController, nil
 		case arm.Subtype:
-			return fakearm.NewArmIK(ctx, config.Component{Name: "arm"}, logger)
+			return fakearm.NewArm(ctx, config.Component{
+				Name:                arm.Subtype.String(),
+				ConvertedAttributes: &fakearm.AttrConfig{ArmModel: xarm.ModelName(6)},
```
Comment (spieswl) [inline services/armremotecontrol/builtin/builtin_test.go:71]: I assume this change is so that this test can use the kinematic model of the xArm6?

**Review by spieswl — APPROVED** [review]
Looks good! I like the simplification down to just `fake`.

File: components/arm/fake/fake.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+		},
+		&AttrConfig{},
+	)
+}
+
+// NewArm returns a new fake arm.
+func NewArm(ctx context.Context, cfg config.Component, logger golog.Logger) (arm.LocalArm, error) {
+	var model referenceframe.Model
+	var err error
+	if cfg.ConvertedAttributes != nil {
```
Comment (biotinker) [inline components/arm/fake/fake.go:60]: Probably we should also allow the use of arbitrary files similar to `wrapper`

File: components/arm/fake/fake.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+// AttrConfig is used for converting config attributes.
+type AttrConfig struct {
+	ArmModel string `json:"arm-model"`
+}
+
+func init() {
+	registry.RegisterComponent(arm.Subtype, ModelName, registry.Component{
+		Constructor: func(ctx context.Context, _ registry.Dependencies, config config.Component, logger golog.Logger) (interface{}, error) {
+			if config.Attributes.Bool("fail_new", false) {
```
Comment (biotinker) [inline components/arm/fake/fake.go:40]: Is this a testing thing that got left in? Doesn't seem to have a purpose other than blowing up

File: components/arm/fake/fake.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf AttrConfig
+			return config.TransformAttributeMapToStruct(&conf, attributes)
+		},
+		&AttrConfig{},
+	)
+}
+
+// NewArm returns a new fake arm.
+func NewArm(ctx context.Context, cfg config.Component, logger golog.Logger) (arm.LocalArm, error) {
```
Comment (biotinker) [inline components/arm/fake/fake.go:57]: I don't think we actually need `context` here, right?

**Review by biotinker — CHANGES_REQUESTED** [review]

File: components/arm/fake/fake.go (hunk 5 above)
Comment (raybjork) [reply components/arm/fake/fake.go:57]: I see now we don't, this is an artifact from the old version that I didn't remove

File: components/arm/fake/fake.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	var model referenceframe.Model
+	var err error
+	if cfg.ConvertedAttributes != nil {
+		switch cfg.ConvertedAttributes.(*AttrConfig).ArmModel {
+		case xarm.ModelName(6):
+			model, err = xarm.Model(6, cfg.Name)
+		case xarm.ModelName(7):
```
[the remark this answers is not in the document]
Comment (raybjork) [reply components/arm/fake/fake.go:64]: Good point.  I'll make the change

File: components/arm/fake/fake.go (hunk 4 above)
Comment (raybjork) [reply components/arm/fake/fake.go:40]: Yes this was left in from the code I merged together. I'll take it out.

File: services/armremotecontrol/builtin/builtin_test.go (hunk 2 above)
Comment (raybjork) [reply services/armremotecontrol/builtin/builtin_test.go]: Yes, before it was using another fake kinematic model, which no longer exists

File: robot/framesystem/service_test.go (hunk 1 above)
Comment (raybjork) [reply robot/framesystem/service_test.go:149]: Unfortunately not really, this was all hand calculated.  But you can see the kinematic information in `fake_model.json`.  Does this answer your question?

File: components/arm/fake/fake.go (hunk 3 above)
Comment (raybjork) [reply components/arm/fake/fake.go:68]: yes probably, but this is work that I don't think is necessary now, so I would like to push it to another PR

File: Makefile [hunk 7]
```diff
@@ -52,8 +52,11 @@ lint-go: tool-install
 	export pkgs="`go list $(GO_BUILD_TAGS) -f '{{.Dir}}' ./... | grep -v /proto/`" && echo "$$pkgs" | xargs go vet $(GO_BUILD_TAGS) -vettool=$(TOOL_BIN)/combined
 	GOGC=50 $(TOOL_BIN)/golangci-lint run $(LINT_BUILD_TAGS) -v --fix --config=./etc/.golangci.yaml
 
-lint-web: buf-web
-	cd web/frontend && npm ci --audit=false && npm run rollup && npm run lint
+lint-web:
```
Comment (biotinker) [inline Makefile:55]: Why are we updating the Makefile as part of this? It looks like this branch needs to be rebased off `main` maybe?

File: components/arm/fake/fake.go [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+// ModelName is the string used to refer to the fake arm model.
+const ModelName = "fake"
+
+//go:embed fake_model.json
+var fakeModelJSON []byte
+
+// AttrConfig is used for converting config attributes.
+type AttrConfig struct {
+	FailNew  bool   `json:"fail_new"`
```
Comment (biotinker) [inline components/arm/fake/fake.go:34]: I think this should also be removed, no?

File: components/arm/fake/fake.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+	ArmModel string `json:"arm-model"`
+}
+
+// Validate ensures all parts of the config are valid.
+func (config *AttrConfig) Validate(path string) error {
+	if config.FailNew {
+		return errors.New("whoops")
```
Comment (biotinker) [inline components/arm/fake/fake.go:41]: This too

File: components/arm/fake/fake.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+			return nil, errors.Errorf("fake arm cannot be created, unsupported arm_model: %s", cfg.ConvertedAttributes.(*AttrConfig).ArmModel)
+		}
+	} else {
+		model, err = referenceframe.UnmarshalModelJSON(fakeModelJSON, cfg.Name)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	mp, err := motionplan.NewCBiRRTMotionPlanner(model, 4, logger)
```
Comment (biotinker) [inline components/arm/fake/fake.go:92]: Can we take this opportunity to `const` the default core count for planning?

**Review by biotinker — CHANGES_REQUESTED** [review]
Just a couple small changes, and one thing to think about, other than that looks good

File: components/arm/fake/fake.go (hunk 8 above)
Comment (raybjork) [reply components/arm/fake/fake.go:34]: This is still being used in some tests, so I wasn't inclined to take it out.  Thoughts?

File: components/arm/fake/fake.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+// NewArm returns a new fake arm.
+func NewArm(cfg config.Component, logger golog.Logger) (arm.LocalArm, error) {
+	var model referenceframe.Model
+	var err error
+	if cfg.ConvertedAttributes != nil {
+		switch cfg.ConvertedAttributes.(*AttrConfig).ArmModel {
+		case xarm.ModelName6DOF:
+			model, err = xarm.Model(6, cfg.Name)
```
[the remark this answers is not in the document]
Comment (raybjork) [reply components/arm/fake/fake.go:71]: I think theres room for improvement here for sure, it doesn't sit right with me either that there is this massive switch statement we will keep needing to add to.  But I don't think theres a way to build an arm model from a config name alone, so maybe its ok for now?

I'll switch the arguments though

File: components/arm/fake/fake.go (hunk 10 above)
Comment (raybjork) [reply components/arm/fake/fake.go:92]: Yeah I'll do this

**Review by biotinker — APPROVED** [review]
LGTM

**Pull request merged**